properties/test_index_manipulation.py

```diff
@@ -0,0 +1,273 @@
+import itertools
+
+import numpy as np
+import pytest
+
+import xarray as xr
+from xarray import Dataset
+from xarray.testing import _assert_internal_invariants
+
+pytest.importorskip("hypothesis")
+pytestmark = pytest.mark.slow_hypothesis
+
+import hypothesis.extra.numpy as npst
+import hypothesis.strategies as st
+from hypothesis import note, settings
+from hypothesis.stateful import (
+    RuleBasedStateMachine,
+    initialize,
+    invariant,
+    precondition,
+    rule,
+)
+
+import xarray.testing.strategies as xrst
+
+
+@st.composite
+def unique(draw, strategy):
+    # https://stackoverflow.com/questions/73737073/create-hypothesis-strategy-that-returns-unique-values
+    seen = draw(st.shared(st.builds(set), key="key-for-unique-elems"))
+    return draw(
+        strategy.filter(lambda x: x not in seen).map(lambda x: seen.add(x) or x)
+    )
+
+
+# Share to ensure we get unique names on each draw,
+# so we don't try to add two variables with the same name
+# or stack to a dimension with a name that already exists in the Dataset.
+UNIQUE_NAME = unique(strategy=xrst.names())
+DIM_NAME = xrst.dimension_names(name_strategy=UNIQUE_NAME, min_dims=1, max_dims=1)
+index_variables = st.builds(
+    xr.Variable,
+    data=npst.arrays(
+        dtype=xrst.pandas_index_dtypes(),
+        shape=npst.array_shapes(min_dims=1, max_dims=1),
+        elements=dict(allow_nan=False, allow_infinity=False, allow_subnormal=False),
+        unique=True,
+    ),
+    dims=DIM_NAME,
+    attrs=xrst.attrs(),
+)
+
+
+def add_dim_coord_and_data_var(ds, var):
+    (name,) = var.dims
+    # dim coord
+    ds[name] = var
+    # non-dim coord of same size; this allows renaming
+    ds[name + "_"] = var
+
+
+class DatasetStateMachine(RuleBasedStateMachine):
+    # Can't use bundles because we'd need pre-conditions on consumes(bundle)
+    # indexed_dims = Bundle("indexed_dims")
+    # multi_indexed_dims = Bundle("multi_indexed_dims")
```

Comment:
I think https://github.com/HypothesisWorks/hypothesis/issues/3944 will enable this, though I wouldn't delay this PR to wait.